### PR TITLE
Feature: Entity / relationship versioning

### LINKED ISSUE #7937: Create a new Version of Entities
In #2844, we supported basic Item Versioning (similar to 6.x) in 7.x.

However, this feature does not fully support Configurable Entities (which are special types of Items). Ideally, this feature should allow for Versioning to work more seamlessly for Configurable Entities.  The Item versioning in #2844 would work at a basic level for Entities (as they are Items), but it ignores existing Relationships.  This behavior would be problematic when versioning an Entity that has many Relationships.

Specifically, to fully support Configurable Entities, the Versioning feature MUST:
* Ensure the new version retains the same Entity type (should happen automatically as it is stored in `dspace.entity.type`)
* Decide how the new Version should deal with existing Relationships:
    * Should Relationships be duplicated to the new Version?  (Note: this could result in, for example, one Person Entity linking to two versions of the same Publication Entity.)
    * Should Relationships only exist on the latest Version of an Entity?  (i.e. they are copied to the new version and removed from the old version, thus ensuring no duplicative relationships)
    * Should multiple options be presented to the user who is creating the new version?

Corresponding UI ticket DSpace/dspace-angular#1312

### PR DESCRIPTION
**Pull request by benbosman**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #7937
* Related to https://github.com/DSpace/dspace-angular/issues/1312
* Related to https://github.com/DSpace/dspace-angular/pull/1624

## Description
This introduces support for versioning of entities with relationships. The solution ensures the new version receives the same relationships, and the old version's relationships remain intact even if the new version is altered

A detailed description can be found at https://docs.google.com/document/d/1E7_zUse5ojI_N9hks53GS3NGboXKkxkF6ud1lPmcDP8/edit#heading=h.mqgcspv7tsrj

This is WIP, the discovery part is not ready yet

## Instructions for Reviewers
When testing this functionality, it's best to use the Angular PR https://github.com/DSpace/dspace-angular/pull/1624
There are many automated tests as well.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon. => Most of the code is tests
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java:137]: Remind me again why we unarchive previous versions of Items?  Is this just so they are not returned in search/browse & also to ensure they aren't processed by Curation Tasks or similar?  (We might want to add a brief note here to describe why we do this. I think there's a good reason, and I may have stumbled on why, but it might be good to document it.)

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java:295]: Small note.. While this isn't "wrong", there's an easier way. I think you can just drop the `String.format` and turn this into:
`log.info("Item with uuid {}, handle {} ... ", latestItem.getId(), latestItem.getHandle(), ...);`

But, like I said, this code isn't wrong... it's just slightly more complex than it needs to be.  So, you can either correct these or leave them as-is.  This same is true for all your other log4j code that calls `String.format()`.... this isn't needed anymore for log4j v2.

**Review by tdonohue — COMMENTED** [review]
@benbosman : I haven't done a full review yet, but I wanted to give you some feedback (as I see no one else has made it to this yet).  Overall, the code here seems reasonable.  I'm only around 1/2 way through a review...but I have a few minor questions/comments (nothing major though).

I haven't had a chance to test this yet with the UI.

File: dspace-api/src/main/java/org/dspace/content/Relationship.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/Relationship.java:249]: From publication X's point of view, the latest project it was ever related to is project Y
The creation of project Y2 happened later, and at that point in time publication X was already an older version, and should therefor not be adjusted (remains linked to project Y)

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java:137]: We just tried to avoid all changes we can avoid. This comes from https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java#L68

We're already dealing with a large amount of unrelated bug fixes in this PR, so we'd prefer to stay away from changing this behavior and risking errors

**Comment by benbosman** [comment]
Thanks @tdonohue for your review.
I've now also merged the bug fix in to ensure the older versions are indexed in solr (and ensuring they're not returned in the search). This was a problem that blocked us from supporting the discovery relations on item pages (e.g. the publications on a person page)

File: dspace/config/spring/api/discovery.xml
Comment (abollini) [inline dspace/config/spring/api/discovery.xml:477]: I'm a bit hesitant here. We are forcing users to add this filter to all their configuration but this is mostly a business requirement than a configuration. Would it better to deal with it similarly to what we do with the withdrawn or discoverable flag? as a parameter in the SolrService methods and/or a SearchIndexPlugin?

**Review by abollini — COMMENTED** [review]
hi @benbosman thanks for the hard work here. As anticipated in the last dev mtg I have completed a first round of review over the code. You will find some comments inline. I plan to make a full test of the features running locally also the angular part over the next week. Maybe looking to the result I will note something else but I hope it is useful for you to start to see my feedback earlier rather than later.

File: dspace-api/src/main/java/org/dspace/content/Relationship.java
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/content/Relationship.java:249]: this comment from @benbosman solve my question if I understood correctly :) https://github.com/DSpace/DSpace/pull/8265#discussion_r871174242

**Comment by benbosman** [comment]
I've merged the last bit of functionality in this branch
This adds the `.latestForDiscovery` virtual metadata documented at https://docs.google.com/document/d/1E7_zUse5ojI_N9hks53GS3NGboXKkxkF6ud1lPmcDP8/edit#
They were first named `.latest`

I did also update the PDF and video visualizing the `.latestForDiscovery` virtual metadata with more details

**Review by tdonohue — CHANGES_REQUESTED** [review]
@benbosman : Overall, this looks great & it works.  I've found some minor things inline below.

I also did want to note a few general things we need to think about doing:
1. First, we need to update the REST Contract to remove the note about Entities not having versioning support: See WARNING here https://github.com/DSpace/RestContract/blob/main/version.md
2. It seems like we might need to update Configurable Entities documentation with details about requiring a `relation.*.latestForDiscovery` field to be created for every `relation.*` field?  Is that correct?
    * I'm a little worried about migrations/upgrades for anyone who has created a custom Entity. Do we need to add notes to the upgrade guide about creating these `relation.*.latestForDiscovery` metadata fields? Or is there any way to "workaround the issue" (log a warning?) if the field doesn't exist?  I expect someone will forget to create a `latestForDiscovery` field in their registry at some point, and I don't know how DSpace will react.

While I understand the reason for this design, I have to admit I'm slightly worried about what will happen if someone tries to misuse any `relation.*` field to store unexpected metadata. We already have hit that same issue with `dspace.entity.type` being misused (see https://github.com/DSpace/dspace-angular/issues/1654).  Just noting this because I'm wondering if we should create a ticket for future work to restrict editing of "relation.*" fields?  (UPDATE: Tested this a bit, overall we are looking good)

Overall, I approve of the code here.  I'm just wanting to point out questions I have around how to ensure simple human mistakes don't cause this feature to break entirely.  There are too many mistakes I see with people misusing or accidentally deleting metadata fields, and I want to be sure it won't break their site if they do so. 

UPDATE: I have tested the following:
* Misused a `relation.isAuthorOfPublication` field by adding a person name. (Nothing obvious broken. Item still works)
* Misused a `relation.​isAuthorOfPublication.​latestForDiscovery` field by adding a name (again, nothing obvious broken. Item still works)
* Attempted to delete both fields from an Item that had them. Was blocked from doing so (likely cause they are virtual fields)
* I haven't yet tested what happens if the corresponding `latestForDiscovery` field just doesn't exist in the Metadata Registry.

**Comment by benbosman** [comment]
@tdonohue I agree we should at some point restrict what metadata can be added/removed/edited. We did also encounter this in the ORCID work, where the owner is also just a metadata field.

I think this would be a very good added value, but quite beyond the scope of this PR. We did already have to include quite some other unrelated improvements

If the metadata field doesn't exist in the registry, I believe the following happens (based on an earlier test):
* The log files will spawn errors or warnings about the missing field
* The discovery indexing for those fields won't work, so they can't be used for searching
* But nothing major goes wrong

I will check all the inline feedback in detail as well

**Comment by mwoodiupui** [comment]
I would suggest, **for a future patch**, adding an "internal use" Boolean attribute to metadatafieldregistry, which tells the UI to disable changes to metadata values for that field and tells dspace-api to reject such changes.  IIRC we already have a few other fields that DSpace expects to manage on its own.

File: dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceIT.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceIT.java:618]: The Service method didn't work reliably, so we have replaced it. We now directly support adding a relationship at the end. This is now done using the create method, with place `-1`. There are tests for that method instead

File: dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceIT.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceIT.java:635]: The Service method didn't work reliably, so we have replaced it. We now directly support adding a relationship at the end. This is now done using the create method, with place `-1`. There are tests for that method instead

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java:2766]: The reason to test `isProjectOfPerson` and `isAuthorOfPublication` separately is to ensure we test this when useForPlace is enabled and disabled. The behavior is quite different in those cases, and did need quite some attention

Good point about the tests of different types of relationships, we will work on that

File: dspace-api/src/test/java/org/dspace/content/service/ItemServiceTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/service/ItemServiceTest.java:83]: We can rewrite this test now, the entity type from the collection had quite some issues when using versioning, but we've solved them in this PR as well
We will adjust the test, as it proves this works correctly now

File: dspace/config/spring/api/discovery.xml
Comment (benbosman) [reply dspace/config/spring/api/discovery.xml:477]: Summary of what we discussed in the meeting:
* I was also originally looking for a good solution, and did weigh a few options
* Because this PR had to deal with many unrelated/unexpected issues, we didn't want a larger change to discovery as well
* The hardcoded solution similar to the workflow doesn't work here because it's not a single configuration (you can have as many as you want, and define them how you'd want)
* The alternative I had in mind was a special DiscoveryConfiguration for this purpose, which doesn't filter on `latestVersion`. The default DiscoveryConfiguration would filter on `latestVersion`. Although this is possible, I would want to keep the scope still limited since this is possible via configuration

The change does indeed require discovery re-indexing. When 7.3 is released, the many DB changes will of course already start a discovery re-index

File: dspace-api/src/main/java/org/dspace/content/RelationshipMetadataServiceImpl.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/RelationshipMetadataServiceImpl.java:112]: This has always been the consequence of tilted items.
The only difference is that we're now using the `latestForDiscovery` virtual metadata as well, which also had to follow the same performance improvements (to ensure we still support tilted items with good performance)

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java:372]: This is indeed completely the same, so we'll refactor that

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java:346]: The `RelationshipMetadataService.getEntityTypeFromMetadata(context,item)` returns an `EntityType` object and requires a context (currently not present in all calls to this method), but if you'd prefer we can check whether it can be refactored

**Comment by benbosman** [comment]
Thanks for all the great feedback, I've planned processing for next week

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java:346]: @benbosman : Honestly, I'll leave it up to you which is easier.  I just noticed the same logic duplicated in both areas.  If it's easy enough to create/pass a Context to this method, then we may want to refactor (to avoid duplicate code).  If that's too complex though, I'm OK with this minor duplication for now.

File: dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/versioning/VersioningConsumer.java]: This has been updated where it was not causing complexities

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: This has been fixed

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java:1818]: The comments have been added

File: dspace/config/modules/versioning.cfg
Comment (benbosman) [reply dspace/config/modules/versioning.cfg]: This has been removed now

File: dspace-api/src/main/java/org/dspace/content/Relationship.java
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/Relationship.java:258]: That's correct

File: dspace-api/src/test/java/org/dspace/content/service/ItemServiceTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/service/ItemServiceTest.java:104]: The additional tests have been created

File: dspace-api/src/test/java/org/dspace/util/RelationshipVersioningUtils.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/util/RelationshipVersioningUtils.java]: This has been renamed

File: dspace-api/src/test/java/org/dspace/content/VersioningWithRelationshipsTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/VersioningWithRelationshipsTest.java]: The tests have been added

File: dspace-api/src/test/java/org/dspace/content/VersioningWithRelationshipsTest.java
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/VersioningWithRelationshipsTest.java]: The tests have been added

File: dspace/config/registries/relationship-formats.xml
Comment (tdonohue) [inline dspace/config/registries/relationship-formats.xml:15]: While I understand what you are trying to say in this `scope_note` (and others similarly worded), I find this description very confusing.  I think the "point of view" phrase is just confusing things.

I wonder if we should reword this to be something like:
> Contains all AUTHOR uuids that the current PUBLICATION links to via a relationship. In other words, this stores all relationships pointing from the current PUBLICATION to an AUTHOR. Internally used by DSpace. Do not manually add, remove or edit values.

Then, for the corresponding `latestForDiscovery` field, we could reword its scope note to say something like:
> Contains all AUTHOR uuids of AUTHORs which link to the current PUBLICATION via a "latest" relationship. In other words, this stores all relationships pointing to the current PUBLICATION from an AUTHOR. Internally used by DSpace to support versioning. Do not manually add, remove or edit values.

OR can `latestForDiscovery` be more easily described as the "inverse relationship"?  For example
> Contains all AUTHOR uuids of AUTHORs which link to this PUBLICATION via inverse relationship (isPublicationOfAuthor). Internally used by DSpace to support versioning....

Are those descriptions accurate?  I just feel the terminology "from the point of view of" is very confusing...as I don't understand what is meant by "point of view".  I'm worried it just won't be understandable to others looking at these scope notes.

**Review by tdonohue — COMMENTED** [review]
@benbosman : Thanks for the updates here!  I've given this a code review and my prior feedback has been addressed, except for the following:
1. We need a small PR to update the REST Contract.  It's necessary to remove the WARNING at https://github.com/DSpace/RestContract/blob/main/version.md
2. We need to update the Configurable Entities documentation to clarify how to "upgrade" to support versioning.  There needs to be additions about the `latestForDiscovery` fields (I'm assuming we need to require they are created for every corresponding `relation.*` field).   I know we've had sites already create custom Entities, so they need to know how to update them to support Versioning.  https://wiki.lyrasis.org/display/DSDOC7x/Configurable+Entities

Both of those can be addressed after this PR is merged, but I want to keep highlighting them as I don't want them to be forgotten.

I have one other minor comment inline with possible suggestions for improving the `scope_note` descriptions. 

I haven't had a chance to re-test this yet... but I recall my last tests succeeded. I'll try to retest this with the frontend PR tomorrow just to verify no new bugs appeared after the code updates

**Comment by benbosman** [comment]
Thanks @tdonohue 

I will create a PR for the contract, and will update the documentation

I did merge another change in from Bruno from our testing. I did notice that if you'd re-index discovery, you're missing the older version items in discovery. This should be fixed now (at least it works correctly for me now)

**Review by tdonohue — COMMENTED** [review]
@benbosman : As noted in the frontend PR, I tested this again today.  Overall it works **great**!  The only thing I discovered was a performance issue in creating a new version of an Entity that has many relationships.  As noted in my [other review](https://github.com/DSpace/dspace-angular/pull/1624#pullrequestreview-997216922), if you attempt to version "Simmons, Cameron" (from the demo entity data set), it takes around 30 seconds for the new version to be created... this seems to be because this Person Entity is linked to 190 Publication Entities, so it's a larger number of Publications to update.

As noted in my frontend review, for now, I think we could "workaround" this issue by providing a message to users that versioning "may take some time".  That said, longer term, we may need to figure out if there's any way to improve the performance when larger numbers of Relationships are involved.

Overall, my prior feedback has been addressed (except for formal Documentation in the wiki, but I do understand that will be coming).

I'm basically a +1, but I did want to give us a moment to discuss what best to do regarding this performance issue.  I'm OK with creating a future ticket for this work, provided we can provide a notice in the UI, and we may want to add a warning to the documentation as well.  Thanks!

File: dspace/config/registries/relationship-formats.xml
Comment (benbosman) [reply dspace/config/registries/relationship-formats.xml]: This has been merged now

**Comment by benbosman** [comment]
Hi @tdonohue we've analyzed the performance in detail. Most of the time is spent on determining the places of the relationships. It's possible to optimize this, but this is the most complicated part of the PR, and I don't think it's wise to redesign this so close to the release.
We've added a warning in the UI about this, and will create a ticket to optimize this afterwards (when there's enough time to thoroughly test the changes as well).
We'll also work on the documentation, I see the wiki is up again

**Review by tdonohue — APPROVED** [review]
👍 Thanks @benbosman !  All my prior feedback has been addressed, and I agree to moving the performance issues to the separate ticket #8352.  I retested today, and it's all still working well.  I like that the UI now warns the user that the version creation process may take some time.

So, in my opinion, this is a good initial implementation of Entities/Relationship versioning.  Future improvements can be made (including performance improvements) in 7.4 (or later).

**Comment by tdonohue** [comment]
@abollini : This PR is waiting on a re-review from you.  All [your prior feedback](https://github.com/DSpace/DSpace/pull/8265#pullrequestreview-969749368) looks to have been addressed (I just double checked). Thanks!

**Review by abollini — APPROVED** [review]
Thanks @benbosman for addressing my feedback. I'm ok in merging this as is in 7.3 and postpone and move the latestVersion filter, as by your last point in comment https://github.com/DSpace/DSpace/pull/8265/files#r877895760 ,  outside of the configuration in a future PR

**Comment by tdonohue** [comment]
@benbosman : Apologies but after merging the ORCID auth PR, this PR (and the UI PR) has new merge conflicts.  This should be the next major feature to be merged if you can get it updated tomorrow.

**Comment by benbosman** [comment]
The merge conflicts are solved, this was agreed to be merged

**Pull request merged**
